src/main/scala/io/archivesunleashed/app/WriteGraph.scala

```diff
@@ -0,0 +1,239 @@
+      case Some(x) => x._2
+      case None => -1
+    }
+  }
+
+  /**  Produces uniqueIds for edge lists from a flatMapped rdd, using nodesWithIds
+   * @param rdd an RDD of elements in format ((datestring, source, target), count)
+   * @return an RDD in format (date, sourceid, destinationid, count)
+   */
+  def edgeNodes (rdd: RDD[((String, String, String), Int)]) : RDD[(String, Long, Long, Int)] = {
+    val nodes = nodesWithIds(rdd)
+    rdd.map(r => (r._1._2, (r._1._3, r._1._1, r._2)))
+       .rightOuterJoin(nodes) // create ids for source urls from nodes.
+       .filter(r => r._2._1 != None)
+       .map( { case (source, (Some((destination, date, weight)), sid))
+         => (destination, (sid, source, date, weight)) })
+      .leftOuterJoin(nodes) // create ids for destination urls from nodes.
+      .filter( r => r._2._1 != None)
+      .map ({ case (dest, ((sid, source, date, weight), Some(did)))
+        => (date, sid, did, weight)})
+  }
+
+  /** Produces the GEXF output from a RDD of tuples and outputs it to gexfPath.
+   *
+   * @param rdd a RDD of elements in format ((datestring, source, target), count)
+   * @param gexfPath output file
+   * @return true on success.
+   */
+  def asGexf (rdd: RDD[((String, String, String), Int)], gexfPath: String): Boolean = {
+    if (gexfPath.isEmpty()) {
+      false
+    } else
+    {
+      val nodes = nodesWithIds(rdd)
+      val outFile = Files.newBufferedWriter(Paths.get(gexfPath), StandardCharsets.UTF_8)
+      val edges = edgeNodes(rdd).map({ case (date, sid, did, weight) =>
+        edgeStart  +
+        sid + targetChunk +
+        did + "\" weight=\"" + weight +
+        directedChunk +
+        "<attvalues>\n" +
+        "<attvalue for=\"0\" value=\"" + date + endAttribute +
+        "</attvalues>\n" +
+        edgeEnd})
+      outFile.write(gexfHeader +
+        "<graph mode=\"static\" defaultedgetype=\"directed\">\n" +
+        "<attributes class=\"edge\">\n" +
+        "  <attribute id=\"0\" title=\"crawlDate\" type=\"string\" />\n" +
+        "</attributes>\n" +
+        "<nodes>\n")
+      nodes.map(r => nodeStart +
+        r._2 + labelStart +
+        r._1 + endAttribute).collect
+        .foreach(r => outFile.write(r))
+      outFile.write("</nodes>\n<edges>\n")
+      edges.collect.foreach(r => outFile.write(r))
+      outFile.write("</edges>\n</graph>\n</gexf>")
+      outFile.close()
+      true
+    }
+  }
+
+  /** Produces the GEXF output from an Array[Row] and outputs it to gexfPath.
+    *
+    * @param data a Dataset[Row] of elements in format (datestring, source, target, count)
+    * @param gexfPath output file
+    * @return true on success.
+    */
+  def asGexf (data: Array[Row], gexfPath: String): Boolean = {
+    if (gexfPath.isEmpty()) {
+      false
+    } else
+    {
+      val outFile = Files.newBufferedWriter(Paths.get(gexfPath), StandardCharsets.UTF_8)
+      val vertices = scala.collection.mutable.Set[String]()
+      data foreach { d =>
```

Comment:
lets use dot notations to be consistent with other collection operations? same below.